### PR TITLE
fix(graders): treat .agent.md tools: as allow-list not expectation (#586)

### LINKED ISSUE #586: bug: `.agent.md` `tools:` auto-grader does not validate an allow-list
## Summary

Waza documents a custom agent's `.agent.md` `tools:` frontmatter as its allowed tool set and says an implicit grader validates that only those tools are used. The current implicit `agent_tools_implicit` grader does not implement allow-list semantics.

It translates every declared tool into `expect_tools`. That means each listed tool must be observed at least once, while extra undeclared tool calls are allowed. For example, an agent declaring `tools: [read, readFile]` may call `bash` or `web_fetch` and still pass if it also calls `read` and `readFile`. Conversely, an agent that correctly uses only `readFile` can fail merely because it did not need `read`.

This contradicts the documented meaning of `.agent.md tools:` and the custom-agent feature's stated goal of validating that the agent only used declared tools.

## Expected semantics

| Agent frontmatter | Validation behavior |
| --- | --- |
| `tools` omitted | No implicit tool-policy validation. |
| `tools: []` | Every model-initiated tool use is a policy violation. |
| `tools: [read, readFile]` | `read` and `readFile` are permitted; every other tool use is a policy violation. Declared tools are not required to be used. |

Tool-name matching must be exact/canonical by default. It must not treat a frontmatter tool name as a regular expression.

## Proposed changes

- Preserve whether `tools:` was omitted versus explicitly present and empty when parsing `.agent.md` frontmatter.
- Replace the auto-injected `ExpectTools` behavior with a dedicated allow-list form in `tool_constraint` (for example, `allow_only`) or an equivalent explicit policy representation.
- Keep `expect_tools` as a separate, opt-in assertion meaning “this tool must be used.”
- Make the automatic validator inspect every observed tool attempt/event and fail on an undeclared tool.
- Report the declared policy, canonical allowed names, and violations in `results.json` and session logs.
- Update the custom-agent documentation so it accurately distinguishes tool-policy allow-list validation from required-tool assertions.

## Acceptance criteria

- [ ] Omitted `tools:` is distinguishable from explicit `tools: []`.
- [ ] The implicit agent grader does not require each declared tool to be used.
- [ ] An observed built-in, MCP, or subagent tool outside a non-empty allow-list fails validation.
- [ ] An observed tool fails validation when `tools: []` is explicitly declared.
- [ ] Exact/canonical matching, including any supported aliases, is documented and unit tested.
- [ ] Existing explicit `tool_constraint` graders retain their documented behavior and continue to override implicit agent validation where appropriate.
- [ ] Docs and the custom-agent example reflect the corrected semantics.

## Related

- #225 introduced custom-agent evaluation and the implicit grader.
- #544 covers broader optional Copilot-native sandboxing; it is complementary but does not correct this allow-list validation defect.
- A companion runtime-enforcement issue will ensure undeclared tools are rejected before execution, rather than only detected afterward.

### PR DESCRIPTION
**Pull request by spboyer**
Fixes #586.

## Problem

`augmentGradersFromAgent` translated each `.agent.md` `tools:` entry into an `expect_tools` entry on the auto-injected `tool_constraint` grader. That meant:

- **Wrong direction.** The agent had to *exercise* every declared tool to pass. A declaration like `tools: [bash, edit, view]` failed any session that only used `bash` and `edit`.
- **No enforcement.** Undeclared tools were silently allowed. An agent declaring `tools: [view]` could shell out via `bash` and still pass.

Both are the opposite of the `.agent.md` contract: `tools:` is the *permission surface*, an upper bound.

## Fix

- **New field:** `AllowOnly *[]ToolSpecParameters` on `ToolConstraintGraderParameters`. Every observed tool call must match at least one entry; declared entries do **not** need to be exercised. Empty list denies all tool use.
- **Exact match, not regex, for `tool` in `allow_only`.** `bash` in the allow-list will not accidentally match `bashful` — issue #586 calls this out explicitly. Qualifier fields (`command_pattern`, `path_pattern`, `skill_pattern`, `args`) remain regex-based so a hand-written `allow_only` can still scope permissible invocations.
- **Tri-state `AgentFrontmatter.Tools` (`*[]string`).** Distinguishes:
  - `nil` (`tools:` absent) → no grader injected
  - `&[]string{}` (`tools: []`) → deny-all allow-list injected
  - populated → allow-list of those names
- **`augmentGradersFromAgent`** now nil-checks the pointer and populates `AllowOnly`, preserving the existing "skip if the user declared their own `tool_constraint` grader" behavior.
- **Preserves** explicit `expect_tools` / `reject_tools` semantics — they still mean "must have used" / "must not have used". `allow_only` composes with both.

## Scoring

Each observed tool call = one check; each unmatched call = one failure. A session with zero tool calls records one vacuous pass so the check count reflects the policy without a divide-by-zero.

## Not in this PR

- **#585 runtime permission enforcement** stays out of scope. This PR only fixes the grader that scores what happened after the fact.
- **#584** was not touched.

## Verification

```
go test -count=1 ./...           # all packages pass
go vet ./...                     # clean
golangci-lint run ./...          # 0 issues
gofmt -s -w .                    # no diff
```

New tests cover:
- allow-list pass (all calls allowed) and undeclared-tool fail
- "not every declared tool need be used" (the core regression)
- exact-match, not regex (`bash` does NOT match `bashful`)
- case-insensitivity
- deny-all with empty list (call → fail, no calls → vacuous pass)
- `command_pattern` qualifier still restricts allowed invocations
- composition with `expect_tools` + `reject_tools`
- tri-state parsing of `tools:` in `.agent.md` frontmatter
- `augmentGradersFromAgent` injects `AllowOnly` for populated and empty `tools:`, and nothing for absent `tools:`

## Docs / schemas

- `docs/graders/tool_constraint.md` — documents `allow_only` alongside expect/reject, calls out exact-vs-regex semantics.
- `site/src/content/docs/guides/graders.mdx` — updates the option table and examples; corrects prior string-list examples to the actual structured form the grader accepts.
- `site/src/content/docs/guides/custom-agents.mdx` — updates the "After" auto-injection example from the (non-existent) `allow:` key to the real `allow_only:` field.
- `schemas/eval.schema.json`, `schemas/task.schema.json` — add `allow_only` to the `anyOf` and to `properties`.

## Files changed

```
docs/graders/tool_constraint.md
internal/graders/tool_constraint_grader.go
internal/graders/tool_constraint_grader_test.go
internal/models/grader_params.go
internal/orchestration/agent_graders.go
internal/orchestration/agent_graders_test.go
internal/skill/agent.go
internal/skill/agent_test.go
schemas/eval.schema.json
schemas/task.schema.json
site/src/content/docs/guides/custom-agents.mdx
site/src/content/docs/guides/graders.mdx
```

Closes #586

### CODE REVIEW

**Review by spboyer — COMMENTED** [review]
The allow-list implementation is close, but explicit `allow_only` configurations remain unusable because model validation rejects them before grader construction.

Issues to address:
- internal/models/grader_params.go:161 - model validation still requires `expect_tools` or `reject_tools`, rejecting `allow_only`-only configurations

File: internal/models/grader_params.go [hunk 1]
```diff
@@ -132,6 +132,33 @@ type ToolSpecParameters struct {
...
+	//   - non-nil empty slice (`&[]ToolSpecParameters{}`): deny-all — any
+	//     observed tool call is a violation.
+	//   - non-nil populated: only observed calls whose tool name matches an
+	//     entry (exact / case-insensitive on the Tool field, with optional
+	//     pattern qualifiers on other fields) are allowed.
+	//
+	// Tool-name matching is exact (case-insensitive) rather than regex, so
+	// entries injected from an .agent.md `tools:` frontmatter list are not
+	// silently reinterpreted as patterns.
+	AllowOnly *[]ToolSpecParameters `yaml:"allow_only,omitempty" json:"allow_only,omitempty"`
```
Comment (spboyer) [inline internal/models/grader_params.go:161]: **[HIGH]** Logic: `AllowOnly` is now a supported config field, but both `GraderConfig.Validate` (`internal/models/spec.go`) and `TestCase.Validate` (`internal/models/testcase.go`) still reject tool constraints unless `expect_tools` or `reject_tools` is non-empty. As a result, a documented `allow_only`-only grader passes decoding/schema checks but fails before `NewToolConstraintGrader` runs. Update both validation branches to accept `params.AllowOnly != nil`, adjust their error messages, and add eval/task validation regression tests that include an explicit empty deny-all list.

File: internal/graders/tool_constraint_grader.go [hunk 2]
```diff
@@ -111,7 +172,10 @@ func (tc *toolConstraintGrader) Grade(ctx context.Context, gradingContext *Conte
 		failures = append(failures, tc.checkExpectTools(session)...)
 		failures = append(failures, tc.checkRejectTools(session)...)
 
-		totalChecks := tc.countTotalChecks()
+		allowFailures, allowChecks := tc.checkAllowOnly(session)
+		failures = append(failures, allowFailures...)
+
+		totalChecks := tc.countTotalChecks() + allowChecks
 		passedChecks := totalChecks - len(failures)
```
[the remark this answers is not in the document]
Comment (spboyer) [reply internal/graders/tool_constraint_grader.go]: Fixed in ec99bd9d — allow-only scoring now counts every violating call while retaining deduplicated feedback.

File: schemas/eval.schema.json [hunk 3]
```diff
@@ -1214,6 +1219,36 @@
...
+              },
+              "skill_pattern": {
+                "type": "string",
+                "description": "Optional regex matched against the skill argument (case-insensitive)."
+              },
+              "path_pattern": {
+                "type": "string",
+                "description": "Optional regex matched against the path argument (case-insensitive)."
+              }
+            }
```
[the remark this answers is not in the document]
Comment (spboyer) [reply schemas/eval.schema.json:1302]: Fixed in ec99bd9d — tool argument matcher schemas now cover expect_tools, reject_tools, and allow_only through a shared definition aligned with runtime validation.

File: schemas/task.schema.json [hunk 4]
```diff
@@ -1050,6 +1055,31 @@
...
+              "command_pattern": {
+                "type": "string"
+              },
+              "skill_pattern": {
+                "type": "string"
+              },
+              "path_pattern": {
+                "type": "string"
+              }
+            }
```
[the remark this answers is not in the document]
Comment (spboyer) [reply schemas/task.schema.json:1133]: Fixed in ec99bd9d — tool argument matcher schemas now cover expect_tools, reject_tools, and allow_only through a shared definition aligned with runtime validation.

File: internal/models/grader_params.go (hunk 1 above)
Comment (spboyer) [reply internal/models/grader_params.go:161]: Fixed in ec99bd9d — both eval-level and task-level validation now accept non-nil allow_only configurations, including deny-all empty lists.

**Comment by spboyer** [comment]
All review feedback is addressed in ec99bd9d. Full Go tests and golangci-lint pass, including new model-validation, schema, and repeated-call scoring regression coverage.
